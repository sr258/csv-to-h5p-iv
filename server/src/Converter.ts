import * as path from 'path';
import { dir } from 'tmp-promise';
import {
    H5PEditor,
    ILibraryStorage,
    IContentStorage,
    ITemporaryFileStorage,
    IKeyValueStorage,
    fsImplementations,
    H5PConfig,
    H5PPlayer
} from '@lumieducation/h5p-server';
import type { DirectoryResult } from 'tmp-promise';
import ReadonlyFileLibraryStorage from './helpers/ReadonlyFileLibraryStorage';

export default class Converter {
    protected constructor() {
        return;
    }

    public contentStorage: IContentStorage;
    public libraryStorage: ILibraryStorage;
    public temporaryStorage: ITemporaryFileStorage;
    public h5pEditor: H5PEditor;
    public h5pPlayer: H5PPlayer;

    protected contentDir: DirectoryResult;
    protected generalPurposeCache: IKeyValueStorage;
    protected h5pConfig: H5PConfig;
    protected temporaryDir: DirectoryResult;

    public static create = async (): Promise<Converter> => {
        const converter = new Converter();
        await converter.init();
        return converter;
    };

    protected init = async () => {
        this.generalPurposeCache = new fsImplementations.InMemoryStorage();
        this.libraryStorage = new ReadonlyFileLibraryStorage(
            path.join(__dirname, '../h5p-libraries')
        );
        this.contentDir = await dir({
            keep: false,
            unsafeCleanup: true
        });
        this.contentStorage = new fsImplementations.FileContentStorage(
            this.contentDir.path
        );
        this.temporaryDir = await dir({
            keep: false,
            unsafeCleanup: true
        });
        this.temporaryStorage =
            new fsImplementations.DirectoryTemporaryFileStorage(
                this.temporaryDir.path
            );
        this.h5pConfig = new H5PConfig(undefined, {
            contentHubEnabled: false,
            contentUserStateSaveInterval: false,
            platformName: 'csv-to-h5p-iv',
            platformVersion: '0.1.0',
            sendUsageStatistics: false,
            setFinishedEnabled: false,
            siteType: 'local',
            uuid: '8de62c47-f335-42f6-909d-2d8f4b7fb7f5'
        });

        this.h5pEditor = new H5PEditor(
            this.generalPurposeCache,
            this.h5pConfig,
            this.libraryStorage,
            this.contentStorage,
            this.temporaryStorage
        );
        this.h5pPlayer = new H5PPlayer(
            this.libraryStorage,
            this.contentStorage,
            this.h5pConfig
        );
    };

    public destroy = async () => {
        await this.contentDir?.cleanup();
        await this.temporaryDir?.cleanup();
    };
}
