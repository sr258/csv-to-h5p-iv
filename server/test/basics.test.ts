import Converter from '../src/Converter';

describe('basic tests', () => {
    it('initializes and can list libraries', async () => {
        // prepare
        const converter = await Converter.create();

        // test
        await expect(
            converter.libraryStorage.getInstalledLibraryNames()
        ).resolves.toMatchSnapshot();
    });

    it('initializes and has functional content storage', async () => {
        // prepare
        const converter = await Converter.create();

        // test
        await expect(
            converter.contentStorage.listContent()
        ).resolves.toMatchObject([]);
    });

    it('initializes and has functional temporary storage', async () => {
        // prepare
        const converter = await Converter.create();

        // test
        await expect(
            converter.temporaryStorage.listFiles()
        ).resolves.toMatchObject([]);
    });
});
