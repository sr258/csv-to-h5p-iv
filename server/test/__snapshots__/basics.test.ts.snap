// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic tests initializes and can list libraries 1`] = `
Array [
  LibraryName {
    "machineName": "Drop",
    "majorVersion": 1,
    "minorVersion": 0,
  },
  LibraryName {
    "machineName": "EmbeddedJS",
    "majorVersion": 1,
    "minorVersion": 0,
  },
  LibraryName {
    "machineName": "FontAwesome",
    "majorVersion": 4,
    "minorVersion": 5,
  },
  LibraryName {
    "machineName": "H5P.AdvancedText",
    "majorVersion": 1,
    "minorVersion": 1,
  },
  LibraryName {
    "machineName": "H5P.Blanks",
    "majorVersion": 1,
    "minorVersion": 14,
  },
  LibraryName {
    "machineName": "H5P.CKEditor",
    "majorVersion": 1,
    "minorVersion": 0,
  },
  LibraryName {
    "machineName": "H5P.DragNBar",
    "majorVersion": 1,
    "minorVersion": 5,
  },
  LibraryName {
    "machineName": "H5P.DragNDrop",
    "majorVersion": 1,
    "minorVersion": 1,
  },
  LibraryName {
    "machineName": "H5P.DragNResize",
    "majorVersion": 1,
    "minorVersion": 2,
  },
  LibraryName {
    "machineName": "H5P.DragQuestion",
    "majorVersion": 1,
    "minorVersion": 14,
  },
  LibraryName {
    "machineName": "H5P.DragText",
    "majorVersion": 1,
    "minorVersion": 10,
  },
  LibraryName {
    "machineName": "H5P.FontIcons",
    "majorVersion": 1,
    "minorVersion": 0,
  },
  LibraryName {
    "machineName": "H5P.FreeTextQuestion",
    "majorVersion": 1,
    "minorVersion": 0,
  },
  LibraryName {
    "machineName": "H5P.GoToQuestion",
    "majorVersion": 1,
    "minorVersion": 3,
  },
  LibraryName {
    "machineName": "H5P.GuidedTour",
    "majorVersion": 1,
    "minorVersion": 0,
  },
  LibraryName {
    "machineName": "H5P.IVHotspot",
    "majorVersion": 1,
    "minorVersion": 2,
  },
  LibraryName {
    "machineName": "H5P.Image",
    "majorVersion": 1,
    "minorVersion": 1,
  },
  LibraryName {
    "machineName": "H5P.InteractiveVideo",
    "majorVersion": 1,
    "minorVersion": 24,
  },
  LibraryName {
    "machineName": "H5P.JoubelUI",
    "majorVersion": 1,
    "minorVersion": 3,
  },
  LibraryName {
    "machineName": "H5P.Link",
    "majorVersion": 1,
    "minorVersion": 3,
  },
  LibraryName {
    "machineName": "H5P.MarkTheWords",
    "majorVersion": 1,
    "minorVersion": 11,
  },
  LibraryName {
    "machineName": "H5P.MultiChoice",
    "majorVersion": 1,
    "minorVersion": 16,
  },
  LibraryName {
    "machineName": "H5P.Nil",
    "majorVersion": 1,
    "minorVersion": 0,
  },
  LibraryName {
    "machineName": "H5P.OpenEndedQuestion",
    "majorVersion": 1,
    "minorVersion": 0,
  },
  LibraryName {
    "machineName": "H5P.Question",
    "majorVersion": 1,
    "minorVersion": 5,
  },
  LibraryName {
    "machineName": "H5P.Questionnaire",
    "majorVersion": 1,
    "minorVersion": 3,
  },
  LibraryName {
    "machineName": "H5P.SimpleMultiChoice",
    "majorVersion": 1,
    "minorVersion": 1,
  },
  LibraryName {
    "machineName": "H5P.SingleChoiceSet",
    "majorVersion": 1,
    "minorVersion": 11,
  },
  LibraryName {
    "machineName": "H5P.SoundJS",
    "majorVersion": 1,
    "minorVersion": 0,
  },
  LibraryName {
    "machineName": "H5P.Summary",
    "majorVersion": 1,
    "minorVersion": 10,
  },
  LibraryName {
    "machineName": "H5P.Table",
    "majorVersion": 1,
    "minorVersion": 1,
  },
  LibraryName {
    "machineName": "H5P.Text",
    "majorVersion": 1,
    "minorVersion": 1,
  },
  LibraryName {
    "machineName": "H5P.TextUtilities",
    "majorVersion": 1,
    "minorVersion": 3,
  },
  LibraryName {
    "machineName": "H5P.Transition",
    "majorVersion": 1,
    "minorVersion": 0,
  },
  LibraryName {
    "machineName": "H5P.TrueFalse",
    "majorVersion": 1,
    "minorVersion": 8,
  },
  LibraryName {
    "machineName": "H5P.Video",
    "majorVersion": 1,
    "minorVersion": 6,
  },
  LibraryName {
    "machineName": "H5PEditor.ColorSelector",
    "majorVersion": 1,
    "minorVersion": 3,
  },
  LibraryName {
    "machineName": "H5PEditor.DragQuestion",
    "majorVersion": 1,
    "minorVersion": 10,
  },
  LibraryName {
    "machineName": "H5PEditor.Duration",
    "majorVersion": 1,
    "minorVersion": 1,
  },
  LibraryName {
    "machineName": "H5PEditor.InteractiveVideo",
    "majorVersion": 1,
    "minorVersion": 24,
  },
  LibraryName {
    "machineName": "H5PEditor.RadioGroup",
    "majorVersion": 1,
    "minorVersion": 1,
  },
  LibraryName {
    "machineName": "H5PEditor.RangeList",
    "majorVersion": 1,
    "minorVersion": 0,
  },
  LibraryName {
    "machineName": "H5PEditor.SelectToggleFields",
    "majorVersion": 1,
    "minorVersion": 1,
  },
  LibraryName {
    "machineName": "H5PEditor.ShowWhen",
    "majorVersion": 1,
    "minorVersion": 0,
  },
  LibraryName {
    "machineName": "H5PEditor.SingleChoiceSetTextualEditor",
    "majorVersion": 1,
    "minorVersion": 0,
  },
  LibraryName {
    "machineName": "H5PEditor.SummaryTextualEditor",
    "majorVersion": 1,
    "minorVersion": 1,
  },
  LibraryName {
    "machineName": "H5PEditor.TableList",
    "majorVersion": 1,
    "minorVersion": 0,
  },
  LibraryName {
    "machineName": "H5PEditor.Timecode",
    "majorVersion": 1,
    "minorVersion": 2,
  },
  LibraryName {
    "machineName": "H5PEditor.UrlField",
    "majorVersion": 1,
    "minorVersion": 2,
  },
  LibraryName {
    "machineName": "H5PEditor.VerticalTabs",
    "majorVersion": 1,
    "minorVersion": 3,
  },
  LibraryName {
    "machineName": "H5PEditor.Wizard",
    "majorVersion": 1,
    "minorVersion": 2,
  },
  LibraryName {
    "machineName": "Shepherd",
    "majorVersion": 1,
    "minorVersion": 0,
  },
  LibraryName {
    "machineName": "Tether",
    "majorVersion": 1,
    "minorVersion": 0,
  },
  LibraryName {
    "machineName": "jQuery.ui",
    "majorVersion": 1,
    "minorVersion": 10,
  },
]
`;
